logging/writer.go

```diff
@@ -1,29 +1,70 @@
 package logging
 
-import "io"
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
 
 // writer is an implementation of io.Writer that writes to a logger.
 type writer struct {
 	l Logger
 	f func(Logger, string)
+
+	mu  sync.Mutex
+	buf strings.Builder
 }
 
 // NewWriter returns an io.Writer that writes content to the given logger.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewWriter(l Logger) io.Writer {
-	return &writer{l, LogString}
+	return &writer{l: l, f: LogString}
 }
 
 // NewDebugWriter returns an io.Writer that writes content to the given logger
 // as debug messages.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewDebugWriter(l Logger) io.Writer {
-	return &writer{l, DebugString}
+	return &writer{l: l, f: DebugString}
 }
 
 func (w *writer) Write(data []byte) (int, error) {
-	w.f(w.l, string(data))
-	return len(data), nil
+	var n int
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	i := bytes.IndexRune(data, '\n')
+	for i != -1 {
+		w.buf.Write(data[:i])
+		w.f(w.l, w.buf.String())
+		n += w.buf.Len()
+		w.buf.Reset()
+
+		data = data[i+1:]
+
+		i = bytes.IndexRune(data, '\n')
+	}
+
+	// Write the remaining data, if any, into the temporary string buffer
+	w.buf.Write(data)
+
+	return n, nil
```

Comment:
According to [io.Writer](https://golang.org/pkg/io/#Writer)

> Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p).

In this case we have consumed the data, even though we haven't flushed it yet, so I wouldn't consider that an error.

```suggestion
	return len(data), nil
```

Reply:
Fair point.

Reply:
Quick question here, we don't write `\n` characters to the logger yet we still count them, should we do something about this?

Reply:
Nope, the API requires that we tell the caller how much of their content we have successfully consumed, it doesn't matter what we do with it.

Reply:
Cool, I had a little change here to capture the initial length of the data (thus retaining variable `n`). Apart from that it's ready for the second round of review.

Reply:
Ahh, of course!